src/cc/mod.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright (C) 2019, Cloudflare, Inc.
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+//     * Redistributions of source code must retain the above copyright notice,
+//       this list of conditions and the following disclaimer.
+//
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+// IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+// THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+use std::time::Instant;
+
+use crate::cc;
+use crate::recovery::Sent;
+
+// Congestion Control constants
+pub const INITIAL_WINDOW_PACKETS: usize = 10;
+
+pub const INITIAL_WINDOW: usize = INITIAL_WINDOW_PACKETS * MAX_DATAGRAM_SIZE;
+
+pub const MINIMUM_WINDOW: usize = 2 * MAX_DATAGRAM_SIZE;
+
+pub const MAX_DATAGRAM_SIZE: usize = 1452;
+
+pub const LOSS_REDUCTION_FACTOR: f64 = 0.5;
+
+// Available CC algorithms.
+// This is defined in include/quiche.h as well.
+// When you add a new CC, update here and include/quiche.h both.
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub enum Algorithm {
+    Reno = 0, // "reno"
+}
+
+impl std::fmt::Debug for dyn CongestionControl {
```

Comment:
Which is fine, but I got the following if I removed L51:

```
error[E0277]: `dyn cc::CongestionControl` doesn't implement `std::fmt::Debug`
   --> src/recovery.rs:552:28
    |
552 |         write!(f, "{:?} ", self.cc)?;
    |                            ^^^^^^^ `dyn cc::CongestionControl` cannot be formatted using `{:?}` because it doesn't implement `std::fmt::Debug`
    |
    = help: the trait `std::fmt::Debug` is not implemented for `dyn cc::CongestionControl`
    = note: required because of the requirements on the impl of `std::fmt::Debug` for `std::boxed::Box<dyn cc::CongestionControl>`
    = note: required by `std::fmt::Debug::fmt`
```

Which seems, recovery.cc is a trait object so it looks for Debug trait on the trait object first.

Also, Debug on Reno is removed (used to use for test, but no more).

Reply:
So I tried to remove L51 block, and I got the following error:
```
error[E0277]: `dyn cc::CongestionControl` doesn't implement `std::fmt::Debug`
   --> src/recovery.rs:552:28
    |
552 |         write!(f, "{:?} ", self.cc)?;
    |                            ^^^^^^^ `dyn cc::CongestionControl` cannot be formatted using `{:?}` because it doesn't implement `std::fmt::Debug`
    |
    = help: the trait `std::fmt::Debug` is not implemented for `dyn cc::CongestionControl`
    = note: required because of the requirements on the impl of `std::fmt::Debug` for `std::boxed::Box<dyn cc::CongestionControl>`
    = note: required by `std::fmt::Debug::fmt`

error: aborting due to previous error
```

Which seems that, what we want is that when `{:?}` is for `recovery.cc` we want to call Debug trait of Reno (actual object), but since this is `dyn CongestonControl` type. so still looking for Debug of that trait object.

Reply:
This is solved. I moved Debug from mod.rs to reno.rs

Reply:
However I still keep ssthresh(), currently it's only required in the testing code. (e.g. testing ssthresh changed after congestion event). I tried to move out of the trait but it didn't work well yesterday, but try to figure it out.

Reply:
ssthresh() removed at 962f379